client/src/stockfish_client.rs

```diff
@@ -0,0 +1,126 @@
+use crate::arrows::{ArrowData, Arrows, ALPHA};
+use crate::stockfish_interface::{run_stockfish, send_command, update_analysis_arrows, Process};
+use async_std::channel::{unbounded, Receiver, Sender};
+use chess::game::Game;
+use chess::moves::Move;
+use dioxus::prelude::*;
+use once_cell::sync::Lazy;
+use regex::Regex;
+
+type Channel = (Sender<()>, Receiver<()>);
+
+pub const MOVES: usize = 10;
+pub const THREADS: usize = 4;
+pub const DEPTH: usize = 40;
+pub const HASH: usize = 256;
+// How much differences in stockfish evaluation affect the alpha of the arrows
+const ALPHA_SENSITIVITY: f64 = 1.0 / 20.0;
+
+static READY_CHANNEL: Lazy<Channel> = Lazy::new(unbounded::<()>);
+
+fn get_info<'a>(output: &'a str, key: &'a str) -> Option<&'a str> {
+    let re = Regex::new(&format!(r"{key} (\S+)")).unwrap();
+    re.captures(output)
+        .and_then(|cap| cap.get(1).map(|m| m.as_str()))
+}
+
+fn inv_sigmoid(x: f64) -> f64 {
+    (x / (1.0 - x)).ln()
+}
+
+fn sigmoid(x: f64) -> f64 {
+    1.0 / (1.0 + (-x).exp())
+}
+
+fn get_eval(output: &str) -> f64 {
+    ALPHA_SENSITIVITY
+        * get_info(output, "score cp")
+            .unwrap()
+            .parse::<f64>()
+            .unwrap()
+}
+
+// Makes it so the arrow for the best move has the default ALPHA value
+fn eval_to_alpha(eval: f64, evals: &[f64]) -> f64 {
+    sigmoid(inv_sigmoid(ALPHA) + eval - evals.iter().max_by(|a, b| a.total_cmp(b)).unwrap())
+}
+
+async fn wait_until_ready(process: &UseRef<Option<Process>>) {
+    send_command(process, "isready").await;
+    READY_CHANNEL.1.recv().await.ok();
+}
+
+pub async fn toggle_stockfish(
+    analyze: UseState<bool>,
+    stockfish_process: UseRef<Option<Process>>,
+    game: UseRef<Game>,
+    arrows: UseRef<Arrows>,
+) {
+    if *analyze.get() {
+        match run_stockfish().await {
+            Ok(process) => {
+                stockfish_process.set(Some(process));
+                arrows.set(Arrows::new(vec![Move::default(); MOVES]));
```

Comment:
I see this being repeated. Can we implement default for `Arrows` and just let it handle doing this somehow?

Reply:
So the `default` is just an empty `Vec`, and this is sort of a way to clear out the arrows while still having the full size. Maybe it could be a `with_size` or something.